The Future of Pre-built Binaries in Nitro-cli

eraykaradag opened the pull request:
*Description of changes:*

I have added a new sub-command under nitro-cli for fetching and listing binaries. The listing functionality currently works only with our default S3 bucket. However, users can fetch binaries from their own S3 bucket or provide a URI containing an archive of binaries. The sub-command will fetch these binaries and store them locally. The URI format is restricted to S3 or HTTP/HTTPS protocols. The archive file must follow a specific structure, including binaries for both arm64 and x86_64 architectures. The provided URI should directly point to the binary archive file. For users wanting to store multiple binaries, the --download-dir argument is available, allowing the fetch-blobs tool to store binaries in their specified location. Usage remains consistent with the existing workflow - binaries can be accessed through environmental variables as before. For private S3 URIs, users need to configure their AWS credentials using AWS tooling.

Examples of how to use:
`nitro-cli fetch-blobs --URI <URI>`
`nitro-cli fetch-blobs --URI <URI> --download-dir <download directory>`
`nitro-cli fetch-blobs --version <version> #this is will fetch from our default s3 bucket`
`nitro-cli fetch-blobs --list`
`nitro-cli fetch-blobs #default way to fetch binaries. It will fetch the latest version we have in our s3 bucket.`

Please let me know if any commit messages need clarification or if you need more information about the implementation.

Commit:
src/lib.rs: Creating subcommand for fetching the binaries

Commit:
src/main.rs: Matching the new subcommand to integrate with the current cli

Commit:
src/common/commands_parser.rs: Created arguments for new  subcommand

Commit:
Added new dependencies

Commit:
src/lib.rs: Mechanism for fetching binaries from s3 storage

Commit:
src/common/mod.rs: added new error result to NitroCliErrorNum

Commit:
src/lib.rs: architecture decider mechanism added

Commit:
src/lib.rs: Changin fetch strategy

In previous strategy my github workflow was different and i was copying binaries one by one from github to s3. now im only copying the tar file which is the same tar file on github releases. With this approach i can generalize and simplfy fetching mechanism

Commit:
src/lib.rs: Fetching mechanism for http and https URIs

Commit:
src/main.rs: Adding fetch-blobs sub-command into nitro-cli

Commit:
src/lib.rs: Implemented unpacking function to unpack enclave-blobs.txz archived file

This function unpacks only the arhitecture specific binaries. Now from user side things a bit easier. Only thing they need to do is pointing the enclave blobs archive and then rest of stuff will be done by nitro-cli. But the archive file should include both architectures and their binaries.

Commit:
src/common/commands_parser.rs: Unit test for testing the args of fetch-blobs sub-command

Commit:
src/lib.rs: Fixing the folder structure

Commit:
fetch-blobs: New mechanism for listing binaries

I added new functionality to list binaries from default s3 bucket. Also i found more idiomatic way to handle arguments when i read the code.

Commit:
fetch-blobs: Allowing users to provide download directory

I created another cmdline argument for users to provide download directory if they want. With this way they can store multiple binaries and use them with environment variable .

Commit:
fetch-blobs: Latest binary fetch mechanism added

Now users can fetch the latest binaries by default or explicitly. If they provide no arguments for  sub-command it will fetch the latest binaries we have in our default S3 bucket.

atanzu commented on src/common/commands_parser.rs:
```diff
@@ -143,7 +143,50 @@ impl BuildEnclavesArgs {
+            /// The URI of users binary storage.
+            uri: String,
+            /// User provided download directory
+            download_dir: Option<String>,
```
nit: `std::path::Path` or `PathBuf` might be a better option.

atanzu commented on src/common/commands_parser.rs:
```diff
@@ -143,7 +143,50 @@ impl BuildEnclavesArgs {
+            download_dir: Option<String>,
+        },
+        ///If user does not provide URI or version it will pull the latest from our default storage.
+        Latest{
```
Imo this option is similar to the previous one, instead of making a separate enum value here, it could be e.g. an empty optional value or a string literal "latest" for the version string of the previous enum item.

atanzu commented on src/lib.rs:
```diff
@@ -47,7 +54,10 @@ pub const CID_TO_CONSOLE_PORT_OFFSET: u32 = 10000;
 const DEFAULT_BLOBS_PATH: &str = "/usr/share/nitro_enclaves/blobs/";
-
+/// Download path to be used to fetch binaries with `fetch-blobs` subcommand.
+const BLOBS_DOWNLOAD_PATH: &str = "/var/lib/nitro-cli/binaries";
```
Potential caveat: user might not have write access to this path.

I'd recommend to use a Tempdir instead: no problems with rw rights, no garbage on the disc.
The only pessimisation is, if the folder already exists and contains the blobs, we might choose to not to download them again. But I'm not sure if this check is feasible.

eraykaradag replied:
Can't we give nitro-cli some permissions or as i understand nitro-cli has some permissions already no?

atanzu replied:
The reason why I'm mentioned it is, if we have options which don't rely on some assumptions and "just work" in all scenarios, then those might be prefered.

atanzu commented on src/lib.rs:
```diff
@@ -572,6 +584,191 @@ pub fn get_file_pcr(path: String, pcr_type: PcrType) -> NitroCliResult<BTreeMap<
+                new_nitro_cli_failure!(&format!("Version mismatch: {:?}", e), NitroCliErrorEnum::BlobFetcherError)
+            }).ok_or_exit_with_errno(None);
+        }
+        FetchBlobsArgs::Latest{download_dir} => {
```
Again, these 2 match arms are pretty same. If you allow to specify "latest" as a version, 2 cases collapse into a single one naturally.

eraykaradag replied:
They can still do it by explicitly using --version latest

atanzu replied:
In this case it seems a bit counter-intuitive that we need 2 match arms which act exactly the same way. so this doubles my point about `Latest` being unneeded.

atanzu commented on src/lib.rs:
```diff
@@ -572,6 +584,191 @@ pub fn get_file_pcr(path: String, pcr_type: PcrType) -> NitroCliResult<BTreeMap<
+
+    match url.scheme() {
+        "s3" => {
+            let bucket = url.host_str().ok_or("No bucket specified").unwrap();
```
I recommend to move code from the match arms into separate functions.

atanzu commented on src/lib.rs:
```diff
@@ -572,6 +584,191 @@ pub fn get_file_pcr(path: String, pcr_type: PcrType) -> NitroCliResult<BTreeMap<
+                }).ok_or_exit_with_errno(None);
+            Ok(())
+        },
+        _ => {
```
please also consider the `file://` case to handle the situation when a URI points to a local file.

eraykaradag replied:
They can still do it that way by using environmental variable don't they?

atanzu replied:
In this case we have 2 different mechanisms to specify paths: environment variables for local files, and URI argument for remote ones. This seems not intuitive, and nothing prevents us from leaving the env option a legacy one, while allowing to use more explicit ways e.g. `file://` URIs.

eraykaradag committed:
Fixing according to feedbacks

eraykaradag committed:
fetch-blobs: Adding name argument to store and use multiple binaries.

With the feedback that i gather i changed the strategy and now users can store multiple binaries and use them with the name they provide.

eraykaradag committed:
build-enclave: Adding new argument to use specific set of pre-built binaries.

With the new argument users can specify the name of the binary set they want and build-enclave tool will use it if they don't provide it will use the default binaries we shipped with nitro-cli

eraykaradag committed:
fetch-blobs: Adding metadata info to binary sets

User may forget which binaries they have or where did they fetched from. Rightnow only source and date info came into my mind but i may add more.

foersleo commented on Cargo.toml:
```diff
@@ -36,7 +36,10 @@ reqwest = { version = "0.11", features = ["json"] }
+futures-util = "0.3"
 lazy_static = "1.4.0"
+prettytable-rs = "0.10"
+byte-unit = "4.0"
```
Why this version? As far as I can tell the latest version we can get through this spec is 2 years old, while there is more recent versions available.

foersleo commented on Cargo.toml:
```diff
@@ -28,6 +28,10 @@ vsock = "0.3"
 sha2 = "0.9.5"
 hex = "0.4"
+url = "2.4"
+aws-config = "0.55"
```
Why this particular version? If my notion of rust dependencies is still correct this limits us to a 2 year old version `0.55.3`, while there is much more current versions available.

See https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#default-requirements

foersleo commented on Cargo.toml:
```diff
@@ -28,6 +28,10 @@ vsock = "0.3"
 hex = "0.4"
+url = "2.4"
+aws-config = "0.55"
+aws-sdk-s3 = "0.28"
```
Same as above.

foersleo reviewed (commented):
First of all, thank you for implementing this. I think from a functional perspective the code is ok. Having this will help us to decouple delivery of our binary blobs from the tool itself.

I have some comments around the organization of the commits and what our plan for merging is.

Do you plan to merge the commit series as is as 20 individual commits, or are we planning to squash all of this together, or is there a plan to re-organize the changes into a low number of commits, each representing a functional unit?

Personally I'd prefer Option 3, but given the necessary effort involved in that endeavor it might not be the best option for this particular patch series.

There is always a balance to strike when it comes to granularity of commits. One example I'd consider too fine grained is commit "[Added new dependencies](https://github.com/eraykaradag/aws-nitro-enclaves-cli/pull/2/commits/da490d74648a8f7241d2c2b6b9e434c5b3b06ef8)". It does not stand on its own. What are we introducing these dependencies for? Why? I think it is way easier to explain these questions if the code is right there in the same commit - it might not be necessary to explain at all in that case and makes it easier to reason around these dependencies for reviewers or yourself when going back to the code in 2 years from now.

In general I think the questions around "Why?" is something that should be part of all commit messages (exceptions may apply for trivial fixups). There is lots of guides on the internet regarding how to write good commit messages, and I highly suggest to give them a read. There is nothing more frustrating than debugging an issue and coming across code from years ago with confusing/missing/outdated/wrong documentation, which includes commit messages. (Only made even more frustrating if the author is one self).

Why am I writing all of this: Please spend some time, trying to read through the commit series with the eyes of a reviewer or yourself in years from now trying to understand why and how we implemented this. Think about what you would love to see in the structure and documentation (commit messages) of the series and keep that in mind for future series. (or if time permits, even go through the exercise to re-structure the series to get some practice in commit re-organization)

eraykaradag committed:
fix: fetch-blobs: Clear directory when fetch fails

I found a flaw that when an error occurs it wasn't clearing directories since it creates a directory before we try to fetch. Also i realized that i was using NitroCliResult's custom ok_or_exit_with_errno function which exits process if error occurs. I made a mistake by using it in lib.rs now i fixed it to propagate the error.

eraykaradag committed:
Feat: fetch-blobs: Now list mechanism can show if you have upstream binaries in your local or not

With using metadata, list command can check whether set of binary exist locally or not.

eraykaradag committed:
test: fetch-blobs: Added Unit Test to cover general functionality of 2nd Project

I found a way to mock http requests but couldn't mock s3 responses. But we can still use CTS testing to test S3 fetching mechanism. I also found a flow when fetching from http uri. I added sanity check if it's a good response or bad.

eraykaradag committed:
fix: fetch-blobs: Changed blobs download path

Since nitro-cli don't have sudo permissions i had to change the download directory for binaries. i found one in users directory named .nitro_cli which it would solve our problem. Also i found a flaw when an error occurs i had to delete the directory aswell but there is a case that we don't allow users to use same name for binaries. When that error occurs i had to filter the errors and avoid clearing the directory.
